lib/msal-node/test/client/ManagedIdentityApplication.spec.ts

```diff
@@ -0,0 +1,326 @@
+                const managedIdentityApplication =
+                    new ManagedIdentityApplication(systemAssignedConfig);
+                expect(managedIdentityApplication.isIMDS()).toBe(true);
+
+                const networkManagedIdentityResult =
+                    await managedIdentityApplication.acquireToken({
+                        resource: MANAGED_IDENTITY_RESOURCE,
+                    });
+                expect(networkManagedIdentityResult.fromCache).toBe(false);
+
+                expect(networkManagedIdentityResult.accessToken).toEqual(
+                    DEFAULT_MANAGED_IDENTITY_AUTHENTICATION_RESULT.accessToken
+                );
+                expect(
+                    managedIdentityApplication.getConfig.managedIdentityId.getId
+                ).toEqual(DEFAULT_MANAGED_IDENTITY_ID);
+                expect(
+                    managedIdentityApplication.getConfig.managedIdentityId
+                        .getIdType
+                ).toEqual(ManagedIdentityIdType.SYSTEM_ASSIGNED);
+                expect(
+                    managedIdentityApplication.getConfig.managedIdentityId
+                        .getIsUserAssignedId
+                ).toEqual(false);
+
+                // wait two seconds for the token to be saved to the cache, before checking the cache for the token
+                // await new Promise<void>((resolve) => {
+                //     let counter: number = 0;
+                //     const interval = setInterval(() => {
+                //         if (counter === 2000) {
+                //             resolve();
+                //             clearInterval(interval);
+                //         } else {
+                //             counter++;
+                //         }
+                //     }, 1); // 1 millisecond
+                // });
+
+                const cacheManagedIdentityResult =
+                    await managedIdentityApplication.acquireToken({
+                        resource: MANAGED_IDENTITY_RESOURCE,
+                    });
+                expect(cacheManagedIdentityResult.fromCache).toBe(true);
+                expect(cacheManagedIdentityResult.accessToken).toEqual(
+                    DEFAULT_MANAGED_IDENTITY_AUTHENTICATION_RESULT.accessToken
+                );
+                expect(
+                    managedIdentityApplication.getConfig.managedIdentityId.getId
+                ).toEqual(DEFAULT_MANAGED_IDENTITY_ID);
+                expect(
+                    managedIdentityApplication.getConfig.managedIdentityId
+                        .getIdType
+                ).toEqual(ManagedIdentityIdType.SYSTEM_ASSIGNED);
+                expect(
+                    managedIdentityApplication.getConfig.managedIdentityId
+                        .getIsUserAssignedId
+                ).toEqual(false);
+            });
+        });
```

Comment:
Please add more complex tests, for example: 

1. Request token for resource R1 for SAMI
2. Request token for resource R2 for SAMI
3. Request token for resource R2 for UAMI

Assert how they are in the cache (i.e. 3 separate tokens)

4. Make requests again and assert everything comes from the cache.

These tests do not need to be written for each individual source, as they mostly test the state of the cache.

Reply:
@bgavrilMS, I'm unable to implement that test you suggested. It requires two different managed identity applications, with two different mocked network responses. However, this is impossible, because the Managed Identity Application's ManagedIdentityClient.identitySource is static. Once created, we don't create it again (your feedback, and this means that it persists across Managed Identity Application instances). So whatever the mocked network response is upon creation... it will be unable to be changed for a different resource (client id).

Reply:
If I comment out the following if statement, then the test I wrote works (based on what you listed above)
```
// if (!ManagedIdentityClient.identitySource) {
        ManagedIdentityClient.identitySource = this.selectManagedIdentitySource(
            this.logger,
            this.nodeStorage,
            this.networkClient,
            this.cryptoProvider
        );
        // }
```